core/src/gql/schema.rs

```diff
@@ -0,0 +1,1114 @@
+use std::collections::BTreeMap;
+use std::fmt::Display;
+use std::sync::Arc;
+
+use crate::dbs::Session;
+use crate::kvs::Datastore;
+use crate::sql::statements::{DefineFieldStatement, SelectStatement};
+use crate::sql::Kind;
+use crate::sql::{self, Table};
+use crate::sql::{Cond, Fields};
+use crate::sql::{Expression, Geometry};
+use crate::sql::{Statement, Thing};
+use async_graphql::dynamic::{Enum, FieldValue, ResolverContext, Type, Union};
+use async_graphql::dynamic::{Field, Interface};
+use async_graphql::dynamic::{FieldFuture, InterfaceField};
+use async_graphql::dynamic::{InputObject, Object};
+use async_graphql::dynamic::{InputValue, Schema};
+use async_graphql::dynamic::{Scalar, TypeRef};
+use async_graphql::indexmap::IndexMap;
+use async_graphql::Name;
+use async_graphql::Value as GqlValue;
+use rust_decimal::prelude::FromPrimitive;
+use rust_decimal::Decimal;
+use serde_json::Number;
+
+use super::error::{resolver_error, GqlError};
+use super::ext::IntoExt;
+use super::ext::ValidatorExt;
+use crate::gql::error::{internal_error, schema_error, type_error};
+use crate::gql::ext::TryAsExt;
+use crate::gql::utils::{get_record, GqlValueUtils};
+use crate::kvs::LockType;
+use crate::kvs::TransactionType;
+use crate::sql::Object as SqlObject;
+use crate::sql::Value as SqlValue;
+
+macro_rules! limit_input {
+	() => {
+		InputValue::new("limit", TypeRef::named(TypeRef::INT))
+	};
+}
+
+macro_rules! start_input {
+	() => {
+		InputValue::new("start", TypeRef::named(TypeRef::INT))
+	};
+}
+
+macro_rules! id_input {
+	() => {
+		InputValue::new("id", TypeRef::named_nn(TypeRef::ID))
+	};
+}
+
+macro_rules! order {
+	(asc, $field:expr) => {{
+		let mut tmp = sql::Order::default();
+		tmp.order = $field.into();
+		tmp.direction = true;
+		tmp
+	}};
+	(desc, $field:expr) => {{
+		let mut tmp = sql::Order::default();
+		tmp.order = $field.into();
+		tmp
+	}};
+}
+
+fn filter_name_from_table(tb_name: impl Display) -> String {
+	format!("_filter_{tb_name}")
+}
+
+pub async fn generate_schema(
+	datastore: &Arc<Datastore>,
+	session: &Session,
+) -> Result<Schema, GqlError> {
+	let kvs = datastore.as_ref();
+	let tx = kvs.transaction(TransactionType::Read, LockType::Optimistic).await?;
+	let ns = session.ns.as_ref().expect("missing ns should have been caught");
+	let db = session.db.as_ref().expect("missing db should have been caught");
+	let tbs = tx.all_tb(ns, db).await?;
+	let mut query = Object::new("Query");
+	let mut types: Vec<Type> = Vec::new();
+
+	trace!(ns, db, ?tbs, "generating schema");
+
+	if tbs.len() == 0 {
+		return Err(schema_error("no tables found in database"));
+	}
+
+	for tb in tbs.iter() {
+		trace!("Adding table: {}", tb.name);
+		let tb_name = tb.name.to_string();
+		let first_tb_name = tb_name.clone();
+		let second_tb_name = tb_name.clone();
```

Comment:
Is it not possible to clone the tb_name inside the moved closures?

Reply:
I think the borrow checker struggles with this situation, and I get lots of lifetime issues if I don't clone, I think it should be helped by async closures though, I'll have another look when those are supported by our msrv